Loki: Allow queries to be made either forward or backward

BaDo2001 opened the pull request:
**What this PR does / why we need it**:
Allow Loki queries to be made either forward or backward.

This can be configured on a per query basis and on a per data source
basis where query configuration overrides data source configuration.
For backwards compatibility the default direction in the data source
configuration is 'BACKWARDS' which is the only direction previously
supported. The data source configuration becomes the default
configuration for new queries created in Explore or via a panel but
the user can change this with a toggle.

**Which issue(s) this PR fixes**:
Addresses the bulk of #20394 but does not touch add paging functionality.

Fixes #20394

**Special notes for your reviewer**:
Continuation of #36379

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=43860) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=43860) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: jdbaldry<br/>:x: BaDo2001<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=43860) it.</sub>

BaDo2001 commented:
I am continuing the work of @jdbaldry from #36379. I see a number of bugs described in this comment: https://github.com/grafana/grafana/pull/36379#issuecomment-873900656 so I will try to solve these first.

Could you please take a look at the original issue and give me a summary of what else needs to be done in order to get this over the line?

Thanks in advance.

ivanahuckova commented:
@BaDo2001 thank you so much for your contribution! We'll have a look. 🙂

ivanahuckova commented on public/app/plugins/datasource/loki/datasource.ts:
```diff
@@ -200,6 +203,7 @@ export class LokiDatasource
     const queryLimit = isMetricsQuery(target.expr) ? options.maxDataPoints : target.maxLines;
     const query = {
       query: target.expr,
+      direction: target.direction,
```
We should probably have the same logic here as in range query
`let direction = target.direction ?? DEFAULT_QUERY_PARAMS.direction`

BaDo2001 replied:
Yes, this seems like a good idea.

ivanahuckova commented on public/app/plugins/datasource/loki/types.ts:
```diff
@@ -28,6 +28,8 @@ export enum LokiQueryType {
   // Stream = 'stream',
 }
 
+export type LokiDirectionType = 'FORWARD' | 'BACKWARD';
```
Do you think it would make sense to change this to enum as `LokiQueryType` is also enum?

BaDo2001 replied:
Yes this makes things more consistent.

ivanahuckova commented on public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx:
```diff
@@ -25,6 +25,7 @@ export function LokiExploreQueryEditor(props: Props) {
       range={range}
       ExtraFieldElement={
         <LokiOptionFields
+          queryDirection={query.direction ?? datasource.direction ?? 'BACKWARD'}
```
I think `queryDirection={query.direction ?? datasource.direction}` should be sufficient as we are always setting direction in datasource and already defaulting to `BACKWARD` there.

BaDo2001 replied:
Nice catch.

ivanahuckova commented on public/app/plugins/datasource/loki/components/AnnotationsQueryEditor.tsx:
```diff
@@ -34,6 +36,7 @@ export const LokiAnnotationsQueryEditor = memo(function LokiAnnotationQueryEdito
         history={[]}
         ExtraFieldElement={
           <LokiOptionFields
+            queryDirection={queryWithRefId.direction ?? datasource.direction ?? 'BACKWARD'}
```
I think queryDirection={query.direction ?? datasource.direction} should be sufficient as we are always setting direction in datasource and already defaulting to BACKWARD there.

BaDo2001 replied:
Nice catch.

ivanahuckova commented on packages/grafana-data/src/types/query.ts:
```diff
@@ -51,6 +53,11 @@ export interface DataQuery {
+  /**
+   * Specify the order in which the returned logs should be displayed initially
+   */
+  defaultLogsSortOrder?: LogsSortOrder;
```
Can't we just get this based on direction? It seems like the same information as sent in `direction`. And then based on `direction` change the order.

BaDo2001 replied:
The problem is that `direction` is a Loki-only thing meaning we would have to do a type-cast to a plugin type inside a core component. I had it originally like that but it seemed wrong. However I am happy to do it if you think its alright.

ivanahuckova commented on public/app/core/logs_model.test.ts:
```diff
@@ -941,10 +941,10 @@ describe('getSeriesProperties()', () => {
-    expect(result.bucketSize).toBe(4);
-    // From time is also aligned to bucketSize (divisible by 4)
-    expect(result.visibleRange).toMatchObject({ from: 8, to: 30 });
+    // Bucketsize 6 gets shortened to 2 because of new visible range is 10ms vs original range being 30ms
```
Do you know why original test fails and why was this changed?

BaDo2001 replied:
The reason why `from: 8` is failing is because this function doesn't operate with a constant margin on the sides and the maths comes out to this.

The `from: 30` is failing because I changed the function to match the end of the timeframe as well as the start.

ivanahuckova commented on public/app/features/explore/Logs.tsx:
```diff
@@ -396,6 +417,7 @@ class UnthemedLogs extends PureComponent<Props, State> {
           </div>
           <LogsNavigation
             logsSortOrder={logsSortOrder}
+            defaultLogsSortOrder={this.getDefaultLogsSortOrder()}
```
Having 2 logsSortOrder seems confusing. Can't we just change initial `logsSortOrder` if the query had `FORWARD` direction?

BaDo2001 replied:
I agree however the problem is the fact that the default being `FORWARD` changes the whole behaviour of the component. So just passing in the current sort order leaves the component with no clue about how to do pagination.

ivanahuckova replied:
I think that there is a big chance that we will implement this also for other logging data sources. @ifrost What do you think about adding something as `hasQueryDirectionSupport` similarly to https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/query.ts#L100 and if data source implements it, we would set initial `logsSortOrder` based on used order?

ivanahuckova reviewed (commented):
Nice job with this! I see that you have tackled most of the feedback from https://github.com/grafana/grafana/pull/36379. 🙌 Left couple of comments below. Also, it would be awesome if you've added tests for new functionalities.  

@nadinevehling This should be probably also reviewed from UX point. There are couple of my observations/thoughts:
1. Should we have DIRECTION as data source level config? 
2. Is it clear for users what is happening/what range they are seeing if the order is flipped? 
Nadine, let me know if you want to pair up on this!

BaDo2001 commented on public/app/plugins/datasource/loki/datasource.ts:
```diff
@@ -259,6 +264,7 @@ export class LokiDatasource
     return {
       ...DEFAULT_QUERY_PARAMS,
       ...range,
+      direction: direction,
```
Thank you.

ifrost commented on public/app/core/logs_model.ts:
```diff
@@ -257,9 +257,10 @@ export function getSeriesProperties(
   // Clamp time range to visible logs otherwise big parts of the graph might look empty
   if (absoluteRange) {
     const earliestTsLogs = sortedRows[0].timeEpochMs;
+    const latestTsLogs = sortedRows[sortedRows.length - 1].timeEpochMs;
```
This is a good feature so log volume issues shouldn't block it. We can add re-loading of the histogram (manual or automatic) or add some sort of message that log volume doesn't contain forward pages.

ivanahuckova reviewed (commented):
Also one more thing - I am worried about scenario where user decides to run 1 query with `FORWARD` and 1 query with `BACKWARD` direction. Which will make pagination unusable. So I am thinking how to handle this case, or if we should disable other direction in all other query rows with tooltip explaining why is this, or if this should be Explore feature and you would be able to set direction only on Explore pane level. 🤔  

@grafana/observability-experience-squad @nadinevehling  what do you think?

jdbaldry committed:
feat: Allow Loki queries to be made either forward or backward.

This can be configured on a per query basis and on a per data source
basis where query configuration overrides data source configuration.
For backwards compatibility the default direction in the data source
configuration is 'BACKWARDS' which is the only direction previously
supported. The data source configuration becomes the default
configuration for new queries created in Explore or via a panel but
the user can change this with a toggle.

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
style: Fix linting

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
style: Remove ' characters from tooltip

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
fix: Update LokiQueryEditor snapshot

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
fix: Add direction to AnnotationsQueryEditor

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

BaDo2001 committed:
Fix: Fix snapshot tests

BaDo2001 committed:
Refactor: Single loki direction type definition

BaDo2001 committed:
Feat: Clamp logs at the end for chart

BaDo2001 committed:
Style: Change let to const

BaDo2001 committed:
Fix: Fix logs in ascending order with loki FORWARD query

BaDo2001 committed:
Refactor: Change loki direction to enum

BaDo2001 committed:
Fix: Fix log navigation in ascending order with loki FORWARD query

BaDo2001 committed:
Fix: Fix loki direction defaulting logic

ivanahuckova commented:
@nadinevehling here is the preview of this functionality. Basically you can choose if your loki query will return logs starting from the `from` part of time range or `to` part of time range. Here are the loki docs: https://grafana.com/docs/loki/latest/api/#get-lokiapiv1query_range. And here is the functionality: 

https://user-images.githubusercontent.com/30407135/150776128-cf0bb7ca-66d8-4049-93c0-3a311b63aab6.mov

ivanahuckova commented:
Hello @BaDo2001, we've had a discussion about this feature with our UX and we've agreed, that before moving with FORWARD queries feature, we need to do UX research and talk to our users to figure out answers to a couple of questions such as:
- Should we have 1 forward-backward switch for all query editors as running 1 backward and 1 forward query breaks the UI? Should we disable any `forward-backward` switch? Which one? 
- Does mixed data sources feature that we are planning to implement influences this?
- Should chosen query direction (forward-backward) influence visualisation order (asc-desc)? Would users expect this? 
- Should this be data source setting? 

Previously, we've gotten a feedback that sometimes our users are confused what is happening with their logs  - what are they seeing - when they interact with histogram, pagination, line limit,... . So we want to be extra careful to make sure we tackle this correctly from the UX point. 

We didn't realize how complex would this feature be before you implemented this. You did A TON OF GREAT work and helped tremendously with this feature. 🧡 But at this point, we need to review the feature during our planning and dedicate UX time for this. I want to thank you so much for all the great work. We really appreciate it! And you really helped us a lot cause we can use your implementation during UX interviews.

BaDo2001 commented:
@ivanahuckova Thanks for the kind words, I really appreciate it. I completely agree, this feature is harder to get right then it first seemed however I would love to see it implemented in the future as my team is very much looking forward to it. ;) Let me know if I can be of any further help. :)

ivanahuckova commented:
I have converted this to draft so it is clear that this PR doesn't need review at the moment. As soon as we have update on this feature, I'll let you know. Thanks again!

ivanahuckova commented:
Closing this PR for now. For testing purposes, we should use `BaDo2001:BaDo2001/loki-query-forwards` branch.

BaDo2001 commented:
Hi @ivanahuckova! Any updates on this? I am happy to continue working on it if there's something on my part to be done.

ivanahuckova commented:
Hello @BaDo2001! Currently our UX team (@niathadgu @mary-sitzenstatter) is doing research on how our users use explore to interact with their loki logs. So we are waiting to find out how to improve current features and how to expand on new feature - such as forward queries.

mary-sitzenstatter commented:
@BaDo2001 - if you're interested in talking to our UX team, please let @niathadgu know and she can set up some time with ya :)

TewWe commented:
Hi @mary-sitzenstatter! May your team have any news on this topic? You advised that we should reach out to @niathadgu, but there is no GitHub user with this username, so we are not sure how to crosscheck if your UX team had any progress on this and who to ask to move ahead. Thanks in advance :)
